src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/ZLibStreamPool.cs

```diff
@@ -0,0 +1,287 @@
+
+            bool restoreFlow = false;
+            try
+            {
+                if (!ExecutionContext.IsFlowSuppressed())
+                {
+                    ExecutionContext.SuppressFlow();
+                    restoreFlow = true;
+                }
+
+                // There is no need to use weak references here, because these pools are kept
+                // for the entire lifetime of the application. Also we reset the timer on each tick,
+                // which prevents the object being rooted forever.
+                _cleaningTimer = new Timer(x => ((ZLibStreamPool)x!).RemoveStaleItems(),
+                    state: this, Timeout.Infinite, Timeout.Infinite);
+            }
+            finally
+            {
+                if (restoreFlow)
+                {
+                    ExecutionContext.RestoreFlow();
+                }
+            }
+        }
+
+        public static ZLibStreamPool GetOrCreate(int windowBits)
+        {
+            Debug.Assert(windowBits >= WebSocketValidate.MinDeflateWindowBits
+                      && windowBits <= WebSocketValidate.MaxDeflateWindowBits);
+
+            int index = windowBits - WebSocketValidate.MinDeflateWindowBits;
+            ref ZLibStreamPool? pool = ref s_pools[index];
+
+            return Volatile.Read(ref pool) ?? EnsureInitialized(windowBits, ref pool);
+
+            static ZLibStreamPool EnsureInitialized(int windowBits, ref ZLibStreamPool? target)
+            {
+                Interlocked.CompareExchange(ref target, new ZLibStreamPool(windowBits, DefaultTimeoutMilliseconds), null);
+
+                Debug.Assert(target != null);
+                return target;
+            }
+        }
+
+        public ZLibStreamHandle GetInflater()
+        {
+            if (TryGet(_inflaters, out ZLibStreamHandle? stream))
+            {
+                return stream;
+            }
+
+            return CreateInflater();
+        }
+
+        public void ReturnInflater(ZLibStreamHandle stream)
+        {
+            if (stream.InflateReset() != ErrorCode.Ok)
+            {
+                stream.Dispose();
+                return;
+            }
+
+            Return(stream, _inflaters);
+        }
+
+        public ZLibStreamHandle GetDeflater()
+        {
+            if (TryGet(_deflaters, out ZLibStreamHandle? stream))
+            {
+                return stream;
+            }
+
+            return CreateDeflater();
+        }
+
+        public void ReturnDeflater(ZLibStreamHandle stream)
+        {
+            if (stream.DeflateReset() != ErrorCode.Ok)
+            {
+                stream.Dispose();
+                return;
+            }
+
+            Return(stream, _deflaters);
+        }
+
+        private void Return(ZLibStreamHandle stream, List<CacheItem> cache)
+        {
+            lock (cache)
+            {
+                cache.Add(new CacheItem(stream));
+
+                if (Interlocked.Increment(ref _activeCount) == 1)
```

Comment:
It seems that `_activeCount` is always modified when `cache` is being locked, so I am not sure if there is value of using `Interlocked` to modify it

Reply:
The reason is because we actually have to cache objects that we modify concurrently - for inflaters and deflaters. I can remove the interlocked, but will need a `_lock` object that is the same for both of them.